Привет еще раз! меня зовут Люман. Сегодня я проверю твой проект.
<br> Дальнейшее общение будет происходить на "ты" если это не вызывает никаких проблем.
<br> Желательно реагировать на красные комментарии ('исправил', 'не понятно как исправить ошибку', ...)
<br> Пожалуйста, не удаляй комментарии ревьюера, так как они повышают качество повторного ревью.

Комментарии будут в <font color='green'>зеленой</font>, <font color='blue'>синей</font> или <font color='red'>красной</font> рамках:


<div class="alert alert-block alert-success">
<b>Успех:</b> Если все сделано отлично
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Если можно немного улучшить
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка:</b> Если требуются исправления. Работа не может быть принята с красными комментариями.
</div>

-------------------

Будет очень хорошо, если ты будешь помечать свои действия следующим образом:
<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> ..
</div>
<div class="alert alert-block alert-warning">
<b>Изменения:</b> Были внесены следующие изменения ...
</div>

<font color='orange' style='font-size:24px; font-weight:bold'>Общее впечатление</font>
* Спасибо за очередную качественную работу, было интересно ее проверять.
- Я оставил некоторые советы надеюсь они будут полезными или интересными.
- Есть небольшие недочеты, но думаю у тебя не составить труда их исправить
- Жду обновленную работу!

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> Люман, привет! Рад снова получить от тебя обратную связь с ценными советами)
</div>


<font color='orange' style='font-size:24px; font-weight:bold'>Общее впечатление[5]</font>
* Спасибо за усердность!
- Было приятно с тобой сотрудничать.
- Недочеты исправлены - работа полностью корректна
- Не буду больше задерживать, продолжай в том же духе.

# Выбор локации для скважины

Допустим, вы работаете в добывающей компании «ГлавРосГосНефть». Нужно решить, где бурить новую скважину.

Вам предоставлены пробы нефти в трёх регионах: в каждом 10 000 месторождений, где измерили качество нефти и объём её запасов. Постройте модель машинного обучения, которая поможет определить регион, где добыча принесёт наибольшую прибыль. Проанализируйте возможную прибыль и риски техникой *Bootstrap.*

Шаги для выбора локации:

- В избранном регионе ищут месторождения, для каждого определяют значения признаков;
- Строят модель и оценивают объём запасов;
- Выбирают месторождения с самым высокими оценками значений. Количество месторождений зависит от бюджета компании и стоимости разработки одной скважины;
- Прибыль равна суммарной прибыли отобранных месторождений.

In [1]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error



<div class="alert alert-block alert-success">
<b>Успех:</b> Импорты  на месте
</div>


## Загрузка и подготовка данных

### Загрузка данных и первичная оценка

In [1]:
data_1 = pd.read_csv('/datasets/geo_data_0.csv')
data_2 = pd.read_csv('/datasets/geo_data_1.csv')
data_3 = pd.read_csv('/datasets/geo_data_2.csv')

NameError: name 'pd' is not defined

In [3]:
display(data_1.info(), data_2.info(), data_3.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null 

None

None

None

In [4]:
display(data_1, data_2.head(), data_3.head())

id        f0        f1        f2     product
0      txEyH  0.705745 -0.497823  1.221170  105.280062
1      2acmU  1.334711 -0.340164  4.365080   73.037750
2      409Wp  1.022732  0.151990  1.419926   85.265647
3      iJLyR -0.032172  0.139033  2.978566  168.620776
4      Xdl7t  1.988431  0.155413  4.751769  154.036647
...      ...       ...       ...       ...         ...
99995  DLsed  0.971957  0.370953  6.075346  110.744026
99996  QKivN  1.392429 -0.382606  1.273912  122.346843
99997  3rnvd  1.029585  0.018787 -1.348308   64.375443
99998  7kl59  0.998163 -0.528582  1.583869   74.040764
99999  1CWhH  1.764754 -0.266417  5.722849  149.633246

[100000 rows x 5 columns]

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

In [5]:
display(data_1['product'].describe(), data_2['product'].describe(), data_3['product'].describe())

count    100000.000000
mean         92.500000
std          44.288691
min           0.000000
25%          56.497507
50%          91.849972
75%         128.564089
max         185.364347
Name: product, dtype: float64

count    100000.000000
mean         68.825000
std          45.944423
min           0.000000
25%          26.953261
50%          57.085625
75%         107.813044
max         137.945408
Name: product, dtype: float64

count    100000.000000
mean         95.000000
std          44.749921
min           0.000000
25%          59.450441
50%          94.925613
75%         130.595027
max         190.029838
Name: product, dtype: float64

Итого, загружены 3 таблицы, пропуски в данных отсутствуют, по формату данных также нет вопросов.\
Видимо где-то опечатка, месторождений не 10 000, а почти по 100 000 в каждом регионе, по крайней мере, если судить по уникальным id.

### Проверка на дубликаты

Проверка на дубли

In [6]:
display(data_1.duplicated().sum(), data_2.duplicated().sum(), data_3.duplicated().sum())

0

0

0

Полностью повторяющихся строк не выявлено. Также будет полезно проверить на дубликаты столбец id:

In [7]:
display(data_1['id'].duplicated().sum(), data_2['id'].duplicated().sum(), data_3['id'].duplicated().sum())

10

4

4

В столбце id выявлено небольшое кол-во дубликатов, на которые можно взглянуть детальнее:

In [8]:
display(data_1[data_1['id'].duplicated()], data_2[data_2['id'].duplicated(keep = False)], data_3[data_3['id'].duplicated(keep = False)])

id        f0        f1         f2     product
7530   HZww2  1.061194 -0.373969  10.430210  158.828695
41724  bxg6G -0.823752  0.546319   3.630479   93.007798
51970  A5aEY -0.180335  0.935548  -2.094773   33.020205
63593  QcMuo  0.635635 -0.473422   0.862670   64.578675
66136  74z30  1.084962 -0.312358   6.990771  127.643327
69163  AGS9W -0.933795  0.116194  -3.655896   19.230453
75715  Tdehs  0.112079  0.430296   3.218993   60.964018
90815  fiKDv  0.049883  0.841313   6.394613  137.346586
92341  TtcGQ  0.110711  1.022689   0.911381  101.318008
97785  bsk9y  0.378429  0.005837   0.160827  160.637302

id         f0         f1        f2     product
1305   LHZR0  11.170835  -1.945066  3.002872   80.859783
2721   bfPNe  -9.494442  -5.463692  4.006042  110.992147
5849   5ltQ6  -3.435401 -12.296043  1.999796   57.085625
41906  LHZR0  -8.989672  -4.286607  2.009139   57.085625
47591  wt4Uk  -9.091098  -8.109279 -0.002314    3.179103
82178  bfPNe  -6.202799  -4.820045  2.995107   84.038886
82873  wt4Uk  10.259972  -9.376355  4.994297  134.766305
84461  5ltQ6  18.213839   2.191999  3.993869  107.813044

id        f0        f1        f2     product
11449  VF7Jo  2.122656 -0.858275  5.746001  181.716817
28039  xCHr8  1.633027  0.368135 -2.378367    6.120525
43233  xCHr8 -0.847066  2.101796  5.597130  184.388641
44378  Vcm5J -1.229484 -2.439204  1.222909  137.968290
45404  KUPhW  0.231846 -1.698941  4.990775   11.716299
49564  VF7Jo -0.883115  0.560537  0.723601  136.233420
55967  KUPhW  1.211150  3.176408  5.543540  132.831802
95090  Vcm5J  2.587702  1.986875  2.482245   92.327572

Значения у повторяющихся столбцов абсолютно разные. При удалении дубликатов оставлю столбцы с наибольшием значением product (логика остальных столбцов неизвестна).

In [9]:
data_1 = data_1.sort_values(by = 'product', ascending = False).drop_duplicates(subset = ['id'], keep = 'first')
data_2 = data_2.sort_values(by = 'product', ascending = False).drop_duplicates(subset = ['id'], keep = 'first')
data_3 = data_3.sort_values(by = 'product', ascending = False).drop_duplicates(subset = ['id'], keep = 'first')
display(data_1['id'].duplicated().sum(), data_2['id'].duplicated().sum(), data_3['id'].duplicated().sum())
display(data_1.shape, data_2.shape, data_3.shape)

0

0

0

(99990, 5)

(99996, 5)

(99996, 5)

Удаляем дубликаты и убеждаемся, что они удалены.


<div class="alert alert-block alert-success">
<b>Успех:</b> Данные осмотрены, но можно было бы еще как минимум изучить корреляции, так как мы планируем использовать линейные модели, а для них это важно. 
</div>


<div class="alert alert-block alert-info">
<b>Совет:</b>  Советую обратить внимание на библиотеки sweetviz и ydata_profiling помогут в проведении более тщательного EDA анализа. Исследовательский анализ можно делать и с помощью ручного вызова функций дефолтных библиотек. Данные библиотеки хороши для максимизации комфорта презентации результатов анализа бизнес-пользователям.  

Очень просты в использоовании, на вход кладется датафрейм: ydata_profiling.ProfileReport(df) 
</div>


<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> у меня почему-то не получается импортировать ни sweetviz, ни ydata_profiling. Наверное, попробую завтра запустить в VS
</div>

## Обучение и проверка модели

### Подготовка выборок

In [10]:
model = LinearRegression()

target_1 = data_1['product']
features_1 = data_1.drop(['id', 'product'] , axis=1)

target_2 = data_2['product']
features_2 = data_2.drop(['id', 'product'] , axis=1)

target_3 = data_3['product']
features_3 = data_3.drop(['id', 'product'] , axis=1)

In [11]:
features_1_train, features_1_valid, target_1_train, target_1_valid = train_test_split(
    features_1, target_1, test_size=0.25, random_state=12345)
features_2_train, features_2_valid, target_2_train, target_2_valid = train_test_split(
    features_2, target_2, test_size=0.25, random_state=12345)
features_3_train, features_3_valid, target_3_train, target_3_valid = train_test_split(
    features_3, target_3, test_size=0.25, random_state=12345)

### Запуск модели

#### Регион 1

In [12]:
model.fit(features_1_train, target_1_train)
predicted_valid_1 = model.predict(features_1_valid)

#### Регион 2

In [13]:
model.fit(features_2_train, target_2_train)
predicted_valid_2 = model.predict(features_2_valid)



<div class="alert alert-block alert-danger">
    

<b>Ошибка:</b>  Опечатка
  
</div>

<div class="alert alert-block alert-warning">
<b>Исправлено:</b> Поправил
</div>

<div class="alert alert-block alert-success">
    
<b>Успех[2]:</b> 👍
</div>


#### Регион 3

In [14]:
model.fit(features_3_train, target_3_train)
predicted_valid_3 = model.predict(features_3_valid)

**Подведем итог работы модели для 3-х регионов:**

In [15]:
print('Регион 1:', 'Предсказанный средний запас сырья -', predicted_valid_1.mean(), 
      '| RMSE -', mean_squared_error(target_1_valid, predicted_valid_1) ** 0.5,
     '| R2 -', r2_score(target_1_valid, predicted_valid_1))
print('Регион 2:', 'Предсказанный средний запас сырья -', predicted_valid_2.mean(), 
      '| RMSE -', mean_squared_error(target_2_valid, predicted_valid_2) ** 0.5,
     '| R2 -', r2_score(target_2_valid, predicted_valid_2))
print('Регион 3:', 'Предсказанный средний запас сырья -', predicted_valid_3.mean(), 
      '| RMSE -', mean_squared_error(target_3_valid, predicted_valid_3) ** 0.5,
     '| R2 -', r2_score(target_3_valid, predicted_valid_3))

Регион 1: Предсказанный средний запас сырья - 92.59577423727681 | RMSE - 37.662476366738964 | R2 - 0.27330224887953236
Регион 2: Предсказанный средний запас сырья - 68.55686699066834 | RMSE - 0.8869609598538721 | R2 - 0.9996249686647039
Регион 3: Предсказанный средний запас сырья - 94.97837180756353 | RMSE - 40.084555496514156 | R2 - 0.19467854753013814


По метрикам видно, что модель более-менее показала результат по региону 1 и, в меньшей мере, по региону 3.\
По региону 2 результат работы модели очень плох.



<div class="alert alert-block alert-info">
<b>Совет: </b>  Можно было бы использовать циклы/функции и избежать дублирования кода. Дублирование кода, часто приводит к глупым ошибкам, которые тяжело отследить.
</div>

## Подготовка к расчёту прибыли

### Сохраним в переменные основные вводные задачи

In [16]:
# Кол-во точек для исследования
SUBJECTS_RES = 500

# Кол-во точек для выбора
SUBJECTS_FIN = 200

# Бюджет на разработку скважин в регионе, рублей
COST = 10_000_000_000

# Доход с единицы продукта (450 р * 1000)
PRICE = 450000


<div class="alert alert-block alert-info">
    
<b>Совет:</b> Константы принято записывать в ВЕРХНЕМ РЕГИСТРЕ =) Читем здесь https://pythonworld.ru/osnovy/pep-8-rukovodstvo-po-napisaniyu-koda-na-python.html#section-29
    
А большие значение можно записывать так: `10_000_000_000`.
</div> 



<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> поправил, PEP-8 добавил в закладки)
</div>

<div class="alert alert-block alert-success">
    
<b>Успех[2]:</b> Правильное решение)
</div>


### Рассчитаем минимальный средний объем сырья на 1 скважину для безубыточной разработки

In [17]:
min_mean_product = COST/SUBJECTS_FIN/PRICE
print('Минимальный средний объем сырья на 1 скважину:', min_mean_product)
print('Средний объем сырья на 1 скважину по регионам:', 'Регион 1 -', target_1.mean(), 
      '| Регион 2 -', target_2.mean(), '| Регион 3 -', target_3.mean())

Минимальный средний объем сырья на 1 скважину: 111.11111111111111
Средний объем сырья на 1 скважину по регионам: Регион 1 - 92.50173489259076 | Регион 2 - 68.82573913718996 | Регион 3 - 95.0013360752838


По расчетам видно, что в средний запас сырья на 1 скважину в каждом регионе ниже среднего минимума для безубыточной разработки. По региону 2 он особенно низкий.\
Однако, также можно увидеть, что верхняя граница ящика (75%) значений по запасам скважен равен 107-130 и есть большое кол-во скважин, подходящих для безубыточной разработки.\
Далее проверим, какой показатель будет у 200 лучших скажин по предсказанию модели.

<div class="alert alert-block alert-success">
<b>Успех:</b> Значение для безубыточной разработки посчитано верно, с выводом согласен  
</div>

## Расчёт прибыли и рисков 

### Функция для расчета прибыли

In [18]:
def total_income (df):
    df = df.sort_values(by = 'predict', ascending = False).head(SUBJECTS_FIN)
    region_income = df['target'].sum() * PRICE
    total_income = region_income - COST
    return total_income

<div class="alert alert-block alert-danger">
<b>Ошибка:</b> Для выбора 200 лучших скважин нужно использовать предсказанные значения. А вот для подсчета прибыли должны использоваться настоящие значения соответствующих ячеек.

</div>

<div class="alert alert-block alert-warning">
<b>Исправлено:</b> Поправил логику расчета прибыли
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка[2]:</b> Поправил, но не совсем как я просил. Нужно делать отбор/сортировку 200 лучших скважин по предсказаниям, а прибыль считать, используя выбранные скважины по таргетам. 
</div>

<div class="alert alert-block alert-warning">
<b>Исправлено[2]:</b> Да, точно, исправил
</div>

<div class="alert alert-block alert-success">
<b>Успех[3]:</b> Теперь логически все правильно
</div>

<div class="alert alert-block alert-danger">
    
<b>Ошибка[3]:</b> но увы  есть 1 неявный недочет: из-за того, что у нас техника бустрап выполняется с возвращением (replace=True), то могут отобраться скважины с одинаковым индексом, и тогда при подстановки предсказаний в таргеты мы получим не 200 скважин, а 200+ и прибыль получится завышенной (можешь вывести длину таргетов после подстановки - и убедишься в этом) 

    
Тут я вижу три решения:
    
- самый быстрый: добавить еще ограничения для таргетов [:SUBJECTS_FIN]
    
- правильный: либо избавиться от дубликатных индексов, либо сбросить индексы для входных данных в начале функции прибыли.
    
- Правильный и лаконичный:  собрать таргеты и предсказания в датафрейм - тогда они и сэмплироваться и сортироватсья будут как единное целое.
</div>



<div class="alert alert-block alert-warning">
<b>Исправлено[3]:</b> Хм, судя по всему, наиболее полезные знания появляются именно на сборном проекте)
    
Попробовал применить метод с датафреймом, вроде как все получилось, прогнозные значения изменились - выросли по 2-м регионам и сильно упали по региону 2, что несколько напрягает - не закралась ли где-то еще какая-нибудь деталь. С другой стороны, если посмотреть с точки зрения бизнеса - такой выхлоп и должен быть от бизнеса (от 20% и выше от инвестиций)
    
Поправил как формулу расчета прибыли, так и бутстреп. Старый код оставил в маркдауне, на всякий случай.
</div>

<div class="alert alert-block alert-danger">
<b>Ответ[4]:</b> Работа с датафрейм в функции прибыли правильная, но ниже ты допустил ошибку:
    
`tar_pred = tar_pred.sample(n=SUBJECTS_RES, replace=True, random_state=state)`
    
Эта строка перезаписала исходный датафрейм, и если на первой итерации сэмплирование проводилось из 25 тысяч скважин, то все последующие уже из 500-ста скважин
    
P.S. Насчет бесполезности данного спринта ты зря так) на самом деле: ведь можно сказать, что весь проект посвящён теме бутстрепа). Совершенно бытовая ситуация - можно ли по одному опыту делать вывод? Например, мы один раз разбили кружку, можно ли по одному случаю утверждать, что мы суперспециалисты по разбиванию кружек? скорее всего нам надо, как минимум, знать а сколько раз мы брали кружку... предпоожим выясняется, что брали её 1000 раз с самой верхней полки и только раз разбили... - не такие уж мы и специалисты по разбиванияю....это один вывод. и т.д.)

ВОт и здесь об этом же: ведь может так случиться, что модель предсказала не совсем верно. Тогда, сделав вывод только по одному замеру 200-от скважин, мы сделаем и ложные управленческие выводы. И тогда вопрос: а как подстраховаться от ложных выводов? И тогда на сцену выходит бутсреп (это что-то вроде моделирования ситуаций на подобие методов монте-карло). Идея бутсрепа такая: взглянем 1000 раз (вместо одно раза) на среднюю из 200-от скважин добычу. Т.е. посчитаем среднее средних.. И уже по нему будем делать вывод. Именно такой подход снизит вероятность сделать ошибочный вывод.


Вообще, методы модеирования (симуляции) ситуаций - распространённый способ управленческого прогнозирования.

</div>

<div class="alert alert-block alert-warning">
<b>Исправлено[4]:</b> Ошибку исправил, в т.ч. скорректировал итоги.

Насчет спринта, я тут скорее не о бесполезности, а о том, что было маловато информации о реальных кейсах и как оно бывает в жизни, какие ситуации и проблемы типичны и т.п. Вот этого лично мне не хватило) 
И рад, что выполняя проект узнал много полезного
</div>


<div class="alert alert-block alert-info">
<b>Совет[5]: </b> Если интересно изучить тему бутстрапа глубже - я оставлю  ниже несколько интересных и полезных ссылок по теме:
        
1. "Bootstrap Hypothesis Testing in Statistics with Example" 
        
https://www.youtube.com/watch?v=9STZ7MxkNVg
        
2. "How to Calculate Bootstrap Confidence Intervals For Machine Learning Results in Python" 
        
https://machinelearningmastery.com/calculate-bootstrap-confidence-intervals-machine-learning-results-python/
        
3. "A Gentle Introduction to the Bootstrap Method" 

https://machinelearningmastery.com/a-gentle-introduction-to-the-bootstrap-method/
        
4. "An Introduction to the Bootstrap Method" 
        
https://towardsdatascience.com/an-introduction-to-the-bootstrap-method-58bcb51b4d60#:~:text=The%20basic%20idea%20of%20bootstrap,population%20mean)%20on%20sample%20data.&amp;text=A%20sample%20from%20population%20with%20sample%20size%20n
        
5. "Bootstrapping and Resampling in Statistics with Example" 
        
        
https://www.youtube.com/watch?v=O_Fj4q8lgmc

6. Также одна из программ на рынке по этой теме управлненеческого моделирования - AnyLogic - интересный продукт. Можешь видео из интернета посмотреть по какому-нибудь примеру с её использованием. Тогда тема управленческого моделирования и бутстрапа может по-другому лечь.
</div>



По расчетам видно, что на предсказанных данных, лучший регион с максимальной прибылью - Регион 1.

### Считаем риски и доверительный интервал с помощью bootstrap

In [19]:
state = np.random.RandomState(12345)

def boot_income(target, predicts):
    predicts = pd.Series(predicts, index = target.index, dtype = 'float64')
    values = []
    for i in range(1000):
        predicts_sample = predicts.sample(n=SUBJECTS_RES, replace=True, random_state=state)
        target_sample = target[predicts_sample.index]
        values.append(total_income(target_sample, predicts_sample))
        check = target.shape
    values = pd.Series(values)
    lower = values.quantile(0.025)
    upper = values.quantile(0.975)
    mean = values.mean()
    risk = (values < 0).sum()/len(values)
  
    return mean, lower, upper, risk, check

<div class="alert alert-block alert-success">
<b>Успех:</b> С бутстрапом  все в порядке (сэмплирование происходит с возвращением), метрики считаются верным образом!  
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка:</b> А в чем смысл в твоей текущей реализации сэмплировать таргеты и подставлять индексы в предсказания?) Разве это не тоже самое, что сэмплировать сразу предсказания
    
P.S. Тебе нужно будет в функцию прибыли передавать и таргеты и предсказания, чтобы отсортировать и отобрать скважины по предсказаниям, а посчитать прибыль по таргетам
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> Действительно, перечитал еще раз описание задачи - все так
</div>

<div class="alert alert-block alert-warning">
<b>Исправлено:</b> Поправил функцию - и результаты стали интереснее и логичнее)
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка[2]:</b> Еще не совсем то, что нужно)

<div class="alert alert-block alert-warning">
<b>Исправлено[2]:</b> Теперь вроде все как надо
</div>

print('Средняя прибыль', ' | Квантиль 2.5%', ' | Квантиль 97.5%', ' | Вероятность убытка')
display(boot_income(target_1_valid, predicted_valid_1))
display(boot_income(target_2_valid, predicted_valid_2))
display(boot_income(target_3_valid, predicted_valid_3))

In [20]:
def boot_income(target, predicts):
    predicts = pd.Series(predicts, index = target.index, dtype = 'float64')
    predicts_df = predicts.to_frame(name = 'predict')
    target_df = target.to_frame(name = 'target')
    tar_pred = pd.concat([target_df, predicts_df], axis = 1)
    values = []
    for i in range(1000):
        tar_pred_sampled = tar_pred.sample(n=SUBJECTS_RES, replace=True, random_state=state)
        values.append(total_income(tar_pred_sampled))
    values = pd.Series(values)
    lower = values.quantile(0.025)
    upper = values.quantile(0.975)
    mean = values.mean()
    risk = (values < 0).sum()/len(values)
  
    return mean, lower, upper, risk

In [21]:
display(boot_income(target_1_valid, predicted_valid_1))
display(boot_income(target_2_valid, predicted_valid_2))
display(boot_income(target_3_valid, predicted_valid_3))

(408808230.7434647, -94004562.24368577, 912504633.4275926, 0.059)

(396531496.83119005, 15337828.978548719, 764569029.1750042, 0.019)

(307798850.0623721, -240056041.12206292, 838515883.1635386, 0.136)

**Код ниже просто для проверки функции**

In [22]:
predictss = pd.Series(predicted_valid_1,index = target_1_valid.index, dtype = 'float64')

predictss_df = predictss.to_frame(name = 'predict')
target_df = target_1_valid.to_frame(name = 'target')
tabl = pd.concat([target_df, predictss_df], axis = 1)
tabl = tabl.sort_values(by = 'predict', ascending = False).head(SUBJECTS_FIN)
region_income = tabl['target'].sum() * PRICE
display(region_income, tabl['target'].sum(), tabl.shape)

13250721037.279392

29446.046749509762

(200, 2)

Итого, что имеем:

1. Для начала, повторим итоги работы модели:\
Регион 1: Предсказанный средний запас сырья - 92.59577423727681 | RMSE - 37.662476366738964 | R2 - 0.27330224887953236\
Регион 2: Предсказанный средний запас сырья - 68.55686699066834 | RMSE - 0.8869609598538721 | R2 - 0.9996249686647039\
Регион 3: Предсказанный средний запас сырья - 94.97837180756353 | RMSE - 40.084555496514156 | R2 - 0.19467854753013814\

По метрикам можно сделать вывод, что модель по региону 2 срабатывает практически идеально, хоть и средние запасы в нем несколько ниже, чем в других регионах.


2. Теперь обратим внимание на результаты бутстрепа:\
Регион 1: Средняя прибыль: 408 808 230 | Квантиль 2.5%: -94 004 562 | Квантиль 97.5%: 912 504 633 | Риск убытка: 5.9%\
Регион 2: Средняя прибыль: 396 531 496 | Квантиль 2.5%: 15 337 828 | Квантиль 97.5%: 764 569 029 | Риск убытка: 1.9%\
Регион 3: Средняя прибыль: 307 798 850 | Квантиль 2.5%: -240 056 041 | Квантиль 97.5%: 838 515 883 | Риск убытка: 13.6%


Выбор региона.

**Регион 1:** Самая высокая оценка средней прибыли по результатам бутстрепа среди регионов - 409 млн потенциальной прибыли. При этом, доверительный интервал составляет от -94 млн до 912 млн, разброс получился широким. При этом регион показывает непроходной риск убытка - 5.9 %.\
**Регион 2:** Здесь немного ниже прогноз прибыли - 396 млн, при этом менее широкий разброс доверительного интервала (15 млн - 764 млн), а также приемлемый уровень риска убытков - 1.9%\
**Регион 3:** Кроме самых низких прогнозов прибыли, в данном регионе также самый высокий риск убытков - 13.6%, по условиям не проходит.

По итогам прогноза, к разработке можно рекомендовать только Регион 2, по которому модель предсказывает относительно высокую прибыль и низкий уровень риска.




<div class="alert alert-block alert-info">
<b>Совет: </b>
    
Не забудь обновить выводы, если потребуется
</div>         

<div class="alert alert-block alert-warning">
<b>Исправлено:</b> Итоги скорректировал. Правда при каждом запуске результаты функции немного разнятся, но не думаю, что это будет значительно
</div>

<div class="alert alert-block alert-warning">
<b>Исправлено[2]:</b> Итоги дополнил
</div>

<div class="alert alert-block alert-warning">
<b>Исправлено[3]:</b> Итоги поправил в очередной раз
</div>

<div class="alert alert-block alert-success">
<b>Успех[5]:</b> Отлично, проведено детальное исследование и был выбран правильный регион!
</div>





<div class="alert alert-block alert-info">
<b>Совет[2]: </b> Если запускать заново код целиком - то все будет детерминировано, не волнуйся
</div>         

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные подготовлены
- [x]  Выполнен шаг 2: модели обучены и проверены
    - [x]  Данные корректно разбиты на обучающую и валидационную выборки
    - [x]  Модели обучены, предсказания сделаны
    - [x]  Предсказания и правильные ответы на валидационной выборке сохранены
    - [x]  На экране напечатаны результаты
    - [x]  Сделаны выводы
- [x]  Выполнен шаг 3: проведена подготовка к расчёту прибыли
    - [x]  Для всех ключевых значений созданы константы Python
    - [x]  Посчитано минимальное среднее количество продукта в месторождениях региона, достаточное для разработки
    - [x]  По предыдущему пункту сделаны выводы
    - [x]  Написана функция расчёта прибыли
- [x]  Выполнен шаг 4: посчитаны риски и прибыль
    - [x]  Проведена процедура *Bootstrap*
    - [x]  Все параметры бутстрепа соответствуют условию
    - [x]  Найдены все нужные величины
    - [x]  Предложен регион для разработки месторождения
    - [x]  Выбор региона обоснован